docs: add how-to-write-fuzz-tests.md (#3763)
GreptimeTeam/greptimedb: docs/how-to/how-to-write-fuzz-tests.md

This document introduce how to write fuzz tests in GreptimeDB.

> ## What is the fuzz test

Reviewer: I would suggest "What is a fuzz test"

Fuzz tests are tools that leverage deterministic random generation to assist in finding bugs. The goal of fuzz tests is to identify inputs generated by the fuzzer that cause system panics, crashes, or unexpected behaviors to occur.

> 

Reviewer 2: Please adds some reference docs about the fuzz test framework or develop guide.

## Why we need them
- Find bugs leverage random generation
- Integrate with other tests(e.g., e2e)

## Resources
All fuzz test-related resources are located in the `/tests-fuzz` directory.
There are two type of resources:

### Fundamental components 
They are located in the `/tests-fuzz/src` directory. The fundamental components define how to generate SQLs (including dialects for different protocols) and validate execution results (e.g., column attribute validation), etc.

### Test targets
They are located in the `/tests-fuzz/targets` directory, with each file representing an independent fuzz test case. The target utilizes fundamental components to generate SQLs, sends the generated SQLs via specified protocol, and validates the results of SQL execution.

> Figure 1 illustrates that the fundamental components of the fuzz test provide the ability to generate random SQLs. It utilizes a Random Number Generator (Rng) to generate the Intermediate Representation (IR), then employs a DialectTranslator to produce specified dialects for different protocols. Finally, the fuzz tests send the generated SQL via the specified protocol and verify that the execution results meet expectations.

Reviewer: `that` - is not needed : `Figure 1 illustrates the fundamental`

> This document introduce how to write fuzz tests in GreptimeDB.

Reviewer 3: ```suggestion
# How to write fuzz tests?

This document introduces how to write fuzz tests in GreptimeDB.
```

## What is a fuzz test
Fuzz test is tool that leverage deterministic random generation to assist in finding bugs. The goal of fuzz tests is to identify inputs generated by the fuzzer that cause system panics, crashes, or unexpected behaviors to occur. And we are using the [cargo-fuzz](https://github.com/rust-fuzz/cargo-fuzz) to run our fuzz test targets. 

## Why we need them

> - Find bugs leverage random generation
> - Integrate with other tests(e.g., e2e)

Reviewer 3: ```suggestion
- Find bugs by leveraging random generation
- Integrate with other tests (e.g., e2e)
```

## Resources

> All fuzz test-related resources are located in the `/tests-fuzz` directory.

Reviewer 3: ```suggestion
All fuzz test-related resources are located in the `tests-fuzz` directory.
```

                             |                                  
                             |                                  
                             v                                  
               Intermediate representation (IR)                 
                             |                                  
                             |                                  
      +----------------------+----------------------+           
      |                      |                      |           
      v                      v                      v           
MySQLTranslator    PostgreSQLTranslator   OtherDialectTranslator
      |                      |                      |           
      |                      |                      |           
      v                      v                      v           
SQL(MySQL Dialect)         .....                  .....         
      |
      |
      v
  Fuzz Test

```
(Figure1: Overview of fuzz tests)

For more details about fuzz targets and fundamental components, please refer to this [tracking issue](https://github.com/GreptimeTeam/greptimedb/issues/3174).

## How to add a fuzz test target

> 1. Create an empty rust source file under the `/tests-fuzz/targets/<fuzz target name>.rs` directory.

Reviewer 3: ```suggestion
1. Create an empty rust source file under the `/tests-fuzz/targets/<fuzz-target>.rs` directory.
```

May use a shorter and no whitespace placeholder. Ditto below.

Reviewer 3: @WenyXu this should be applied Ditto below.

2. Register the fuzz test target in the `/tests-fuzz/Cargo.toml` file.

```toml
[[bin]]
name = "{fuzz target name}"
path = "targets/<fuzz target name>.rs"
test = false
bench = false
doc = false

> 

Reviewer 3: ```suggestion
```

```

3. Define the `FuzzInput` in the `/tests-fuzz/targets/<fuzz target name>.rs`.

```rust
#![no_main]
use libfuzzer_sys::arbitrary::{Arbitrary, Unstructured};

#[derive(Clone, Debug)]
struct FuzzInput {
    seed: u64,
}

impl Arbitrary<'_> for FuzzInput {
    fn arbitrary(u: &mut Unstructured<'_>) -> arbitrary::Result<Self> {
        let seed = u.int_in_range(u64::MIN..=u64::MAX)?;
        Ok(FuzzInput { seed })
    }
}
```

4. Write your first fuzz test target in the `/tests-fuzz/targets/<fuzz target name>.rs`.

```rust
use libfuzzer_sys::fuzz_target;
use rand::{Rng, SeedableRng};
use rand_chacha::ChaChaRng;
use snafu::ResultExt;
use sqlx::{MySql, Pool};
use tests_fuzz::fake::{
    merge_two_word_map_fn, random_capitalize_map, uppercase_and_keyword_backtick_map,
    MappedGenerator, WordGenerator,
};
use tests_fuzz::generator::create_expr::CreateTableExprGeneratorBuilder;
use tests_fuzz::generator::Generator;
use tests_fuzz::ir::CreateTableExpr;
use tests_fuzz::translator::mysql::create_expr::CreateTableExprTranslator;
use tests_fuzz::translator::DslTranslator;
use tests_fuzz::utils::{init_greptime_connections, Connections};

> ...

Reviewer 3: ```suggestion
```

Reviewer 3: Keep it compilable.

fuzz_target!(|input: FuzzInput| {
    common_telemetry::init_default_ut_logging();
    common_runtime::block_on_write(async {
        let Connections { mysql } = init_greptime_connections().await;
            let mut rng = ChaChaRng::seed_from_u64(input.seed);
            let columns = rng.gen_range(2..30);

>             ...

Reviewer 3: ```suggestion
// ...?
```

Use comment to describe anything that should be added here.

            let create_table_generator = CreateTableExprGeneratorBuilder::default()
                .name_generator(Box::new(MappedGenerator::new(
                    WordGenerator,
                    merge_two_word_map_fn(random_capitalize_map, uppercase_and_keyword_backtick_map),
                )))
                .columns(columns)
                .engine("mito")
                .if_not_exists(if_not_exists)
                .build()
                .unwrap();
            let ir = create_table_generator.generate(&mut rng);
            let translator = CreateTableExprTranslator;
            let sql = translator.translate(&expr).unwrap();

>             ...

Reviewer 3: ```suggestion

```

ditto

            mysql.execute(&sql).await
    })
});
```

5. Run your fuzz test target

> Please follow the instructions in this [document](../../tests-fuzz/README.md).

Reviewer 3: ```suggestion
Please follow the instructions in this [document](/tests-fuzz/README.md).
```

> Please follow the instructions in this [document](../../tests-fuzz/README.md).

Reviewer 3: We can duplicate a happy path here, like:

```
cargo fuzz run fuzz_create_table --fuzz-dir tests-fuzz
```

and add the ref to test-fuzz's README for other more complicated cases.

use tests_fuzz::translator::mysql::create_expr::CreateTableExprTranslator;
use tests_fuzz::translator::DslTranslator;
use tests_fuzz::utils::{init_greptime_connections, Connections};

fuzz_target!(|input: FuzzInput| {
    common_telemetry::init_default_ut_logging();
    common_runtime::block_on_write(async {
        let Connections { mysql } = init_greptime_connections().await;
            let mut rng = ChaChaRng::seed_from_u64(input.seed);
            let columns = rng.gen_range(2..30);
            let create_table_generator = CreateTableExprGeneratorBuilder::default()
                .name_generator(Box::new(MappedGenerator::new(
                    WordGenerator,
                    merge_two_word_map_fn(random_capitalize_map, uppercase_and_keyword_backtick_map),
                )))
                .columns(columns)
                .engine("mito")
                .if_not_exists(if_not_exists)
                .build()
                .unwrap();
            let ir = create_table_generator.generate(&mut rng);
            let translator = CreateTableExprTranslator;
            let sql = translator.translate(&expr).unwrap();
            mysql.execute(&sql).await

>             // other logical

Reviewer 3: ```suggestion
```

No need. Only necessary if it's something MUST DO in the framework but should be completed by users. It's common that users can add any customized logics.

    })
});
```

> 5. Run your fuzz test target
> 
> ```bash
>     cargo fuzz run <fuzz-test> --fuzz-dir tests-fuzz
> ```

Reviewer 3: ```suggestion
5. Run your fuzz test target:

```bash
cargo fuzz run <fuzz-target> --fuzz-dir tests-fuzz
```
```

The placeholder is the same as `fuzz-target`?

Author: Good catch!

This document introduce how to write fuzz tests in GreptimeDB.

> 

Reviewer 2: Let's add some descriptions about fuzz tests, what it is, and why we need them.

All fuzz test-related resources are located in the `/tests-fuzz` directory.
- Each target(located in the `/tests-fuzz/targets`) stands for an independent fuzz test case.
- Fuzz test fundamental components are located in the `/tests-fuzz/src`.

> Figure 1 illustrates that it utilizes a Rng (Random Number Generator) to generate the IR (Intermediate Representation), then employs a DialectTranslator to produce specified dialects for different protocols.

Reviewer 2: Produce specified dialects for different protocols., then what? Looks like missing something.

                             |                                  
                             v                                  
               Intermediate representation (IR)                 
                             |                                  
                             |                                  
      +----------------------+----------------------+           
      |                      |                      |           
      v                      v                      v           
MySQLTranslator    PostgreSQLTranslator   OtherDialectTranslator
      |                      |             |           
      |                      |                      |           
      v                      v                      v           
 MySQL Dialect             .....                  .....         
```
(Figure1: Overview of fuzz test components)

## How to add a fuzz test target

1. Create an empty rust source file under the `/tests-fuzz/targets/` directory.

2. Register the fuzz test target in the `/tests-fuzz/Cargo.toml` file.
```toml
...
[[bin]]
name = "{target name}"
path = "targets/{target}.rs"
test = false
bench = false
doc = false
...
```

> 3. Define the `FuzzInput`

Reviewer 2: Where to put this code?

```rust
#[derive(Clone, Debug)]
struct FuzzInput {
    seed: u64,
}

impl Arbitrary<'_> for FuzzInput {
    fn arbitrary(u: &mut Unstructured<'_>) -> arbitrary::Result<Self> {
        let seed = u.int_in_range(u64::MIN..=u64::MAX)?;
        Ok(FuzzInput { seed })
    }
}
```

> 4. Your first fuzz test target

Reviewer 2: The same question as above.

use tests_fuzz::translator::mysql::create_expr::CreateTableExprTranslator;
use tests_fuzz::translator::DslTranslator;
use tests_fuzz::utils::{init_greptime_connections, Connections};

fuzz_target!(|input: FuzzInput| {
    common_telemetry::init_default_ut_logging();
    common_runtime::block_on_write(async {
        let Connections { mysql } = init_greptime_connections().await;
            let mut rng = ChaChaRng::seed_from_u64(input.seed);
            let columns = rng.gen_range(2..30);

> // ...?

Reviewer 3: @WenyXu This suggestion doesn't for applied directly. Please read the original comment.
